velox/exec/tests/utils/QueryAssertions.cpp

```diff
@@ -703,9 +793,115 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
   return result.first->task();
 }
 
-// Special function to compare multisets with vectors of variants in a way that
-// we compare all floating point values inside using 'epsilon' constant.
-// Returns true if equal.
+// Try to bucket rows in left and right into leftBucketed and rightBucketed
+// and returns whether the bucketization is successful. The bucketization is
+// successful if the following requirements are satisfied.
+//   1. left and right are non-empty and have the same number of columns of the
+// same types.
+//   2. Rows in left and right are grouped by all columns except the
+// floating-point column via sorting. In at least one of them, each group has at
+// most one row.
+// The bucket results are stored in leftBucketed and rightBucketed with the
+// floating-point columns moved to the end and rows in ascending order.
+bool tryBucketMaterializedRows(
```

Comment:
This function groups rows by non-floating-point columns via sorting and if there are multiple floating-point columns, it checks that every group contains at most one row. (I feel the other name I think of is worse than the current one, i.e., "qualifyEpsilonComparison".)

Reply:
> if there are multiple floating-point columns, it checks that every group contains at most one row

Does the check apply only when there are multiple floating point columns. I assume it applies if there is at least one such column, no?

> I feel the other name I think of is worse

Let's try to come up with a few names. Maybe, sortByUniqueKey(<list of columns that are expected to form a unique key>).

Reply:
> Does the check apply only when there are multiple floating point columns.
I'm updating the code to only check every group containing a single row only when there are more than one floating-point columns.

> Maybe, sortByUniqueKey().
This sounds like we already have unique keys and then sort by them. But this function is more like we sort the data to check whether all keys are unique.

Reply:
> I'm updating the code to only check every group containing a single row only when there are more than one floating-point columns.

I'm curious why. It seems it doesn't matter how many floating-point column there are. What's the difference between having 1 or 2 or more?

Reply:
> I feel this sounds like we already have unique keys and then sort by them. But this function is more like we sort the data to check whether all keys are unique.

Another way to look at this is that we expect that a subset of columns forms a unique key and we want to sort by them. In case our assumption is wrong this function will return 'false'.

Reply:
> > I'm updating the code to only check every group containing a single row only when there are more than one floating-point columns.
> 
> I'm curious why. It seems it doesn't matter how many floating-point column there are. What's the difference between having 1 or 2 or more?

We can check every group containing a single row regardless of how many floating-point columns we have. It's just when there is one floating-point column, even if there are multiple rows per group, we can still compare them via sort-merge with epsilon (because even if there are two expected values x < y that are close enough such that their actual approximate x' > y', y' and x' can still match with x and y.) 

If this case makes it hard to understand the behavior of assertEqualResults(), I can just force it to check every group containing one row for the case of one floating-point column too.

Reply:
According to offline discussion, I removed the optimization for the case of one floating-point column. I added a comment about relaxing the condition of single-row-groups under this case in QueryAssertions.h.